state/unit_ops.go

```diff
@@ -57,20 +58,37 @@ func (op *unitSetStateOperation) Build(attempt int) ([]txn.Op, error) {
 			Id:     unitGlobalKey,
 			Assert: txn.DocMissing,
 			Insert: unitStateDoc{
-				DocID: unitGlobalKey,
-				State: escapedState,
+				DocID:         unitGlobalKey,
+				State:         escapedState,
+				UniterState:   op.newState.UniterState,
+				RelationState: op.newState.BSONRelationData(),
+				StorageState:  op.newState.StorageState,
 			},
 		}}, nil
 	}
 
 	// State keys may contain dots or dollar chars which need to be escaped.
-	escapedState := make(bson.M, len(op.newState))
-	for k, v := range op.newState {
+	escapedState := make(bson.M, len(op.newState.State))
+	for k, v := range op.newState.State {
 		escapedState[mgoutils.EscapeKey(k)] = v
 	}
 
 	// Check if we need to update
-	if stDoc.stateMatches(escapedState) {
+	setFields := bson.D{}
+	if !stDoc.stateMatches(escapedState) {
+		setFields = append(setFields, bson.DocElem{"state", escapedState})
+	}
+	if op.newState.UniterState != "" {
+		setFields = append(setFields, bson.DocElem{"uniter-state", op.newState.UniterState})
+	}
+	if len(op.newState.RelationState) > 0 {
```

Comment:
Likewise, if the lengths match we should compare the contents for changes.

Reply:
updated